File: perf/http/metrics.ts
export type Metric = {
    inputTime: Date;
    inputType: string;
    outputs: { outputTime: Date; outputType: string }[];
}

export type OutputMetric = {
    outputTime: Date;
    outputType: string;
}

export class MetricMap {
    private metricMap: Map<string, Metric>;

    constructor() {
        this.metricMap = new Map();
    }

    public newMetric(messageId: string, inputType: string) {
        this.metricMap.set(messageId, {inputTime: new Date(), inputType: inputType, outputs: []});
    }

    appendMetric(messageId: string, outputType: string) {
        const metric = this.metricMap.get(messageId);
        if (!metric) {
            throw new Error(`unknown metric for message id ${messageId}`);
        }
        metric.outputs.push({outputTime: new Date(), outputType});
    }
    
    iterateMetrics(f: (metric: Metric, outputMetric: OutputMetric) => void) {
        for (const [_, metric] of this.metricMap) {
            for (const outputMetric of metric.outputs) {
                f(metric, outputMetric);
            }
        }
    }
}

File: perf/k6/metrics.go
package main

import (
	"log/slog"
	"sync"
	"time"
)

type Metric struct {
	InputTime int64          `js:"inputTime"`
	InputType string         `js:"inputType"`
	Outputs   []OutputMetric `js:"outputs"`
}

type OutputMetric struct {
	OutputTime int64  `js:"outputTime"`
	OutputType string `js:"outputType"`
}

type Metrics struct {
	Lock      sync.Mutex
	MetricMap map[string]Metric
}

func NewMetricMap() *Metrics {
	return &Metrics{
		MetricMap: make(map[string]Metric),
	}
}

func (m *Metrics) NewMetric(messageId string, inputType string) {
	m.Lock.Lock()
	defer m.Lock.Unlock()

	m.MetricMap[messageId] = Metric{
		InputTime: time.Now().UnixMilli(),
		InputType: inputType,
		Outputs:   []OutputMetric{},
	}
}

func (m *Metrics) AppendMetric(messageId string, outputType string) {
	m.Lock.Lock()
	defer m.Lock.Unlock()

	metric, ok := m.MetricMap[messageId]
	if !ok {
		slog.Info("messageId not found in metrics map", "messageId", messageId)
		return
	}
	metric.Outputs = append(metric.Outputs, OutputMetric{
		OutputTime: time.Now().UnixMilli(),
		OutputType: outputType,
	})
	m.MetricMap[messageId] = metric
}
